reactor-core/src/main/java/reactor/core/scheduler/SchedulerMetricDecorator.java

```diff
@@ -0,0 +1,89 @@
+import reactor.core.Scannable;
+import reactor.core.Scannable.Attr;
+
+import static io.micrometer.core.instrument.Metrics.globalRegistry;
+
+final class SchedulerMetricDecorator
+			implements BiFunction<Scheduler, ScheduledExecutorService, ScheduledExecutorService>,
+			           Disposable {
+
+	static final String TAG_SCHEDULER_ID = "reactor.scheduler.id";
+
+	final WeakHashMap<Scheduler, String>        seenSchedulers          = new WeakHashMap<>();
+	final Map<String, AtomicInteger>            schedulerDifferentiator = new HashMap<>();
+	final WeakHashMap<Scheduler, AtomicInteger> executorDifferentiator  = new WeakHashMap<>();
+
+	@Override
+	public synchronized ScheduledExecutorService apply(Scheduler scheduler, ScheduledExecutorService service) {
+		//this is equivalent to `toString`, a detailed name like `parallel("foo", 3)`
+		String schedulerName = Scannable
+				.from(scheduler)
+				.scanOrDefault(Attr.NAME, scheduler.getClass().getName());
+
+		//we hope that each NAME is unique enough, but we'll differentiate by Scheduler
+		String schedulerId =
+				seenSchedulers.computeIfAbsent(scheduler, s -> {
+					int schedulerDifferentiator = this.schedulerDifferentiator
+							.computeIfAbsent(schedulerName, k -> new AtomicInteger(0))
+							.getAndIncrement();
+
+					return (schedulerDifferentiator == 0) ? schedulerName
+							: schedulerName + "#" + schedulerDifferentiator;
+				});
+
+		//we now want an executorId unique to a given scheduler
+		String executorId = schedulerId + "-" +
+				executorDifferentiator.computeIfAbsent(scheduler, key -> new AtomicInteger(0))
+				                      .getAndIncrement();
+
+		/*
+		Design note: we assume that a given Scheduler won't apply the decorator twice to the
+		same ExecutorService. Even though, it would simply create an extraneous meter for
+		that ExecutorService, which we think is not that bad (compared to paying the price
+		upfront of also tracking executors instances to deduplicate). The main goal is to
+		detect Scheduler instances that have already started decorating their executors,
+		in order to avoid consider two calls in a row as duplicates (yet still being able
+		to distinguish between two instances with the same name and configuration).
+		 */
+
+		// TODO return the result of ExecutorServiceMetrics#monitor
+		//  once ScheduledExecutorService gets supported by Micrometer
+		//  See https://github.com/micrometer-metrics/micrometer/issues/1021
+		ExecutorServiceMetrics.monitor(globalRegistry, service, executorId,
+				Tag.of(TAG_SCHEDULER_ID, schedulerId));
+
+		return service;
+	}
+
+	@Override
+	public void dispose() {
+		Search.in(globalRegistry)
+		      .tagKeys(TAG_SCHEDULER_ID)
+		      .meters()
+		      .forEach(globalRegistry::remove);
+
+		this.seenSchedulers.clear();
+		this.schedulerDifferentiator.clear();
+		this.executorDifferentiator.clear();
+	}
+
+	@Override
+	public boolean isDisposed() {
+		return false; //never really disposed
```

Comment:
Don't we break the contract here?
Maybe we should change to `Closeable` instead?

Reply:
nope, the contract says
> Implementations are not required to track disposition and as such may never
> return `true` even when disposed. However, they MUST only return true
> when there's a guarantee the resource or task is disposed.

Reply:
Shouldn't we keep the default implementation from the interface then?

Reply:
we could, yeah. I wanted to be explicit and use that as a chance to state it's something we explicitly thought about, but works either way.